torch/utils/checkpoint.py

```diff
@@ -335,7 +340,370 @@ def forward(input):
+    kwargs = {k: v for k, v in zip(kwarg_keys, flat_args[-len(kwarg_keys) :])}
+    return args, kwargs
+
+class _NoopSaveInputs(torch.autograd.Function):
+    # Autograd Function that saves inputs and returns them as-is
+    # This is not used directly, see _applyAutogradFunctionToSaveInputs below.
+    @staticmethod
+    def forward(*args):
+        return args
+
+    @staticmethod
+    def setup_context(ctx: Any, inputs: Tuple[Any, ...], output: Any) -> None:
+        ctx.save_for_backward(*inputs)
+
+    @staticmethod
+    def backward(ctx, *grad_outputs):
+        return grad_outputs
+
+def _applyAutogradFunctionToSaveInputs(*args: torch.Tensor) -> Tuple[Tuple[torch.Tensor, ...], Node]:
+    # Wrapper around _NoopSaveInputs to preserve the requires_grad-ness of the inputs
+    idx_no_req_grad = [i for i, t in enumerate(args) if isinstance(t, torch.Tensor)
+                       and not t.requires_grad and (t.is_floating_point() or t.is_complex())]
+    new_args = _NoopSaveInputs.apply(*args)
+    return (
+        tuple(t.detach() if i in idx_no_req_grad else t for i, t in enumerate(new_args)),
+        new_args[0].grad_fn
+    )
+
+class _CheckpointFrame():
+    def __init__(self, recompute_fn):
+        self.recompute_fn = recompute_fn
+        self.ctx = None
+
+        # See [ Implementation details of Rule 2 ]
+        self.maybe_args: Optional[Tuple[torch.Tensor, ...]] = None
+        self.weak_holders: List[ReferenceType] = []
+
+        self.recomputed: DefaultDict[int, weakref.WeakKeyDictionary[_Handle, torch.Tensor]] = \
+            defaultdict(weakref.WeakKeyDictionary)
+        self.recomp_counter: DefaultDict[int, int] = defaultdict(lambda: 0)
+        self.is_recomputed: DefaultDict[int, bool] = defaultdict(lambda: False)
+
+# See Rule 5
+class _StopRecomputationError(Exception):
+    pass
+
+class _recomputation_hook(torch.autograd.graph.saved_tensors_hooks):
+    def __init__(self, target_frame_ref: ReferenceType, gid: int):
+        def pack_hook(x):
+            target_frame = target_frame_ref()
+            assert target_frame is not None
+            recomp_idx = target_frame.recomp_counter[gid]
+            target_frame.recomp_counter[gid] += 1
+            holder = target_frame.weak_holders[recomp_idx]()
+
+            if holder is not None:
+                # See Rule 6: [ Multiple backwards ] above
+                if holder.handle is None:
+                    # See Rule 4 above
+                    holder.handle = _Handle()
+                target_frame.recomputed[gid][holder.handle] = x.detach()
+
+            # TODO: figure out why some tests are failing when early stop is not enabled
+            if _enable_checkpoint_early_stop and \
+               target_frame.recomp_counter[gid] == len(target_frame.weak_holders):
+                raise _StopRecomputationError()
+            # See Rule 6: [ Basic case ] above
+            return x.detach()
+
+        def unpack_hook(x):
+            return x
```

Comment:
The original checkpoint had this error because the assumption was that no one should be calling backward on the recomputated graph. But now that we want to support backward within checkpoint, we want the hook active during recomputation to actually do two things 1) I want to save to the weak key dictionary, AND I also want to 2) save to the recomputed graph. There are no lifetime issues with this because whatever is saved on this graph is guaranteed to die when the graph dies.

Also see Rule 6 [Basic case] for more context.

Reply:
This explanation is not very clear to me (not sure what is the link with dicts and saving stuff on the recomputed graph). I still expect that the graph created during the recompute will be discarded straight away and never backwarded on?

Reply:
> (not sure what is the link with dicts and saving stuff on the recomputed graph

yeah no link, but just saying that it needs to do both

>  I still expect that the graph created during the recompute will be discarded straight away and never backwarded on?

No I think we should expect the graph created during recompute to be backwarded on.  Consider the following example:

```python
def fn(x):
   y = x.sin().cos()  # (1)
   gx, = torch.autograd.grad(y, x)  # (2)
   return y.sin()   # (3)

out = checkpoint(fn, a)    # (4)
out.backward()    # (5)
```

Forward:
1. x and x.sin() aren't saved because we're running forward under checkpoint
2. need to recompute fn because we didn't save in (1)
1. this time we aren't running under checkpoint, so we DO save tensors onto the graph (if we don't early stop we proceed to finish running fn, e.g. we would backward on this recomputed graph)
2. now we backward on the original graph (recomputed graph has died)
3. don't save y because we're under checkpoint

Backward
1. Begin recompute, we start building recomputed graph, saving x and x.sin() onto it
2. the recomputed graph already has x and x.sin() saved onto it, so we just backward on the recomputed graph
3. save y onto the recomputed graph
5. actually do backward on the original graph

Reply:
Ho ok, so it is when there is a grad within the fn and:
- During the og fw IF early stop is disabled
- During the og fw IF you have multiple grad, while we recompute things for the second grad
- During the og bw 

Can you leave a comment to that end?